text/1001-deprecate-named-inject.md

```diff
@@ -0,0 +1,186 @@
+      }
+    })
+  
+  // now it's time to replace
+  root.find(j.ClassProperty).forEach(path => {
+    let node = path.node;
+    
+    let hasInject = hasDecorators(node, [...importNames.values()]);
+    
+    if (!hasInject) return;
+    
+    node.decorators = node.decorators.map(decorator => {
+      let { expression } = decorator;
+      
+      if (expression.type === 'Identifier') {
+        if (importNames.has(expression.name)) {
+          decorator.expression = j.identifier('service');
+        }
+      }
+      
+      if (expression.type === 'CallExpression') {
+        decorator.expression.callee = j.identifier('service');
+      }
+      
+      return decorator;
+    });
+  });
+  
+  return root.toSource();
+}
+
+// Copied from: https://github.com/NullVoxPopuli/ember-concurrency-codemods/tree/main
+function firstMatchingDecorator(node, named = []) {
+  if (!node.decorators) return;
+
+  return node.decorators.find((decorator) => {
+    let { expression } = decorator;
+
+    switch (expression.type) {
+      case 'MethodDefinition': {
+      }
+      case 'CallExpression': {
+        let { callee } = expression;
+
+        switch (callee.type) {
+          case 'Identifier':
+            return named.includes(callee.name);
+          case 'MemberExpression': {
+            let { object } = callee;
+
+            return named.includes(object.callee.name);
+          }
+        }
+      }
+      case 'Identifier':
+        return named.includes(expression.name);
+    }
+  });
+}
+
+function hasDecorators(node, named = []) {
+  return Boolean(firstMatchingDecorator(node, named));
+}
+```
+
+<details><summary>The test scenarios</summary>
+
+```ts 
+import { inject } from '@ember/service';
+import { inject as service } from '@ember/service';
+// import Service from '@ember/service';
+import BaseService from '@ember/service';
+import { inject as serviceDecorator } from '@ember/service';
+import { inject as x } from '@ember/service';
+// import { service } from '@ember/service';
+import { service as y } from '@ember/service';
+// import Service, { inject, service } from '@ember/service';
+import Service, { inject as s } from '@ember/service';
+
+
+export default class Demo extends Service {
+  
+}
+
+export default class Demo2 extends BaseService {
+  // simple
+  @inject router;
+  @service router1;
+  @x router2;
+  @y router3;
+  @serviceDecorator router4;
+  @inject('router') router41;
+  
+  // TS-only
+  @inject declare router5: Type;
+  @inject('router') declare router51: Type;
+  @service declare router6: Type;
+  @x declare router7: Type;
+  @y declare router8: Type;
+  @serviceDecorator declare router9: Type;
+}
+```
+
+</detailS>
+
+
+## How We Teach This
+
+The docs / guides already use the new import path.
+
+## Drawbacks
+
+n/a
+
+## Alternatives
+
+do nothing, the cost of an export alias is:
+- a few extra bytes
+- mental gymnastics for teaching
+- "another case to cover" for tooling
```

Comment:
I don't understand this comment, can you elaborate?

Reply:
When tooling tries to implement go-to-definition for services, (because we use string-based services, we have to implement this ourselves), the tool needs to know about all the potential exports (inject and service) in order to look up their usages.

It's not a lot, but it is _one more thing_.

And unifying on _one less thing_ benefits simplicity of tooling implementation, as well as our learning story.